Feature/fat32 write

DrDeano opened the pull request:
Added FAT32 write capabilities. See commits for the journey.
This is a long and painful journey of reverse engineering and the FAT32 manual.
After many months of pain and hard work, this will closes #216

SamTebbs33 commented on mkfat32.zig:
```diff
@@ -523,7 +523,7 @@ pub const Fat32 = struct {
         // Valid clusters are 1, 2, 4, 8, 16, 32, 64 and 128
-        if (options.cluster_size < 0 or options.cluster_size > 128 or !std.math.isPowerOfTwo(options.cluster_size)) {
+        if (options.cluster_size < 1 or options.cluster_size > 128 or !std.math.isPowerOfTwo(options.cluster_size)) {
             log.err("Sectors per cluster is invalid. Must be less then 32 and a power of 2. Found: {}", .{options.cluster_size});
```
One thing I didn't notice before is that the error message says that the cluster size must be less than 32 but the condition uses 128.

DrDeano replied:
Ah yes 👍

SamTebbs33 commented on src/kernel/filesystem/vfs.zig:
```diff
@@ -552,7 +552,13 @@ pub fn isAbsolute(path: []const u8) bool {
+///     Error.NotADirectory - The node isn't a directory node.
+///
+pub fn setRoot(node: *Node) Error!void {
+    if (node.isFile() or node.isSymlink()) {
```
It may be best to check `!isDir()` instead.

DrDeano replied:
Sure

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -441,11 +442,11 @@ const FATConfig = struct {
 
-    /// The number of free clusters on the disk. Will be 0xFFFFFFFF is unknown.
+    /// The number of free clusters on the disk. Will be 0x0FFFFFFF is unknown.
     number_free_clusters: u32,
 
     /// The next free cluster to start looking for when allocating a new cluster for a
-    /// file. Will be 0xFFFFFFFF is unknown.
+    /// file. Will be 0x0FFFFFFF is unknown.
```
It's a nitpick, but the first zeroes in the hex numbers here are unnecessary

DrDeano replied:
Yes, was from a copy and replace.

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -514,6 +515,48 @@ fn initStruct(comptime Type: type, bytes: []const u8) Type {
+/// Arguments:
+///     IN comptime Type: type - The type of the struct.
+///     IN copy_struct: Type   - The struct to copt from.
+///     IN bytes: []u8         - The bytes to copt to.
```
copt -> copy

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -950,7 +1007,7 @@ pub fn Fat32FS(comptime StreamType: type) type {
                     defer if (long_entries) |entries| self.allocator.free(entries);
 
                     // If attribute is 0x0F, then is a long file name and
```
extra "and" here

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -1147,43 +1204,79 @@ pub fn Fat32FS(comptime StreamType: type) type {
-                .CREATE_DIR, .CREATE_FILE, .CREATE_SYMLINK => createFileOrDirOrSymlink(fs, dir, name, flags, open_args),
+                .CREATE_FILE => createFileOrDir(fs, dir, name, false),
+                .CREATE_DIR => createFileOrDir(fs, dir, name, true),
+                // FAT32 doesn't support symlinks
```
I wonder how we could keep symlinks persistent in a disk then 🤔

DrDeano replied:
tbh, FAT32 can get in the bin XD
This is just a starting file system. We will can use Ext2 for our main FS as that has symlinks

SamTebbs33 commented on src/kernel/filesystem/fat32.zig:
```diff
@@ -1277,34 +1377,672 @@ pub fn Fat32FS(comptime StreamType: type) type {
+                }
+            }
+
+            // Update the FAT for the allocated cluster
```
From this point onwards the cluster state will be borked if any of the writes fail but I can't think of a nice way to use an errdefer to fix the state in case that happens. Could a copy of the cluster at its state before the writes be kept around and restored in case an error occurs?

DrDeano replied:
But what if the restore fails as will still need to write to the disk? I think caching might help, but I don't think I can have all the writes happen atomically. Plus, the write will fail if there is an issue with the disk and that would be a bigger problem.

SamTebbs33 replied:
Would one of these writes failing cause any problems for other clusters or the rest of the filesystem? If it would only affect this one cluster/file then I don't see it being corrupted as too big an issue as long as the error gets reported properly so the user knows.

DrDeano replied:
This write would only effect either a newly created file, writing to a file that is bigger than a cluster. so could corrupt the file being worked on and nothing else (I think). The error will bubble up to who called write or open create. It would result in lost clusters, but won't effect the file system as a whole. But depending on which write failed would corrupt the file system in different ways.

SamTebbs33 reviewed (commented):
Nice work :) There are a lot of new tests and I don't have the mental rigidity to read through them all but I trust they are doing the right sort of thing ;)

DrDeano commented:
> 
> 
> Nice work :) There are a lot of new tests and I don't have the mental rigidity to read through them all but I trust they are doing the right sort of thing ;)

I have tried to cover all the normal use cases to ensure everything works as normal, but have probably missed a lot of error cases, but would have too many unit tests. So if there is a error that occurs, can save the corrupt FAT image and use the image as a test case

DrDeano committed:
See description

Added finding the next free cluster
Added name to long name
Added long name to short name
Added tests for above

Added createLongNameEntry + tests

Moved tests to bottom

Added createShortNameEntry + test

Used the RTC for the date and time for the created short entry

Tidied createEntries


A bit of refactor

No symlinks for FAT32

findNextFreeCluster updates FAT

Reordered tests to better follow the FAT32FS code
FAT32 has no support for symlinks, so removed code around this.
Removed open_args from createNode as it doesn't need it

Added writeEntries + tests

Write the short and long entries to disk
findNextFreeCluster update cluster chain with parent cluster

Added FAT32 write + tests

This Added the ability to create files and directories and write to files.
Added location of the short dir entry for the file so can update the size of the file on disk
Added folders to the test FAT32 directory.
Also fixed minor bug in mkfat32

Added check for destroying the filesystem


Fixed error message for cluster size


Simpler if condition


0x0FFFFFFF => 0xFFFFFFFF


Spelling


Fixed test
